### PR TITLE
change search strategy, to match only inside component metadata

### PR DESCRIPTION
**Pull request by cebor**

### CODE REVIEW

**Comment by mirkonasato** [comment]
Only matching inside `@Component` is the first thing I wanted to propose as well.

However the `componentRegex` in this PR doesn't work with nested pairs of curly braces, e.g.

	@Component({
	  selector: 'my-test',
	  providers: [
	    provide(SomeService, {useClass: DefaultSomeService})
	  ],
	  templateUrl: 'test.component.html'
	})

**Comment by cebor** [comment]
so now it should :), made it greedy

**Comment by cebor** [comment]
https://regex101.com/r/pS8dK9/1

**Comment by mirkonasato** [comment]
If it's greedy it can easily match more than just the component metadata: https://regex101.com/r/tQ4cF8/1

That's where things get tricky. :)

**Comment by cebor** [comment]
Oh you are absolutely right, trying to figure this out ...

**Comment by cebor** [comment]
What about this: https://regex101.com/r/tQ4cF8/2

**Comment by mirkonasato** [comment]
I still feel that's too brittle, would match up to wherever you have `})` plus whitespace. To be honest, I think this shows the [limitation of using regexes](http://stackoverflow.com/questions/133601/can-regular-expressions-be-used-to-match-nested-patterns) for this task.

I think we should move this discussion to an issue instead and look at possible approaches there.

**Comment by TheLarkInn** [comment]
I like this approach but maybe want to see an additional test for this before pulling this in. Would you add a new fixture which this code safeguards from?

**Comment by mirkonasato** [comment]
I think if it's not going to be 100% accurate then it should err on the side of replacing too little rather than too much, and document the limitations.

**Comment by TheLarkInn** [comment]
Agreed

**Comment by cebor** [comment]
@TheLarkInn I modified Test5 to cover not replacing `templateUrl` and `styleUrls` outside components metadata

Also the regex should work now properly, when you have no missing semicolons in your component class.

So for example the following will fail: 

```typescript
@Component({
  selector: 'my-test',
  providers: [
    provide(SomeService, {useClass: DefaultSomeService})
  ],
  templateUrl: 'test.component.html'
})
export class TestComponent {
  constructor(someService: SomeService) {
    someService.getSomething({
      option1: 'value1',
      templateUrl: '/this/should/not/match'
    })
    // line above has missing ;  -> regex will fail
  }
}
```

Also this will not work, but i think nobody does it this way:

```typescript
@Component({
  selector: 'my-test',
  providers: [
    provide(SomeService, {useClass: DefaultSomeService})
  ],
  templateUrl: 'test.component.html'
}) export class TestComponent { ... }   // no new line after decorator
```

**Comment by TheLarkInn** [comment]
@cebor Can you rebase changes from master, want to test coverage reports.

File: index.js [hunk 1]
```diff
@@ -1,31 +1,37 @@
 // using: regex, capture groups, and capture group variables.
-var templateUrlRegex = /templateUrl *:(.*)$/gm;
-var stylesRegex = /styleUrls *:(\s*\[[^\]]*?\])/g;
-var stringRegex = /(['"])((?:[^\\]\\\1|.)*?)\1/g;
+const componentRegex = /@Component\({([\s\S]*)}\)$/gm;
```
Comment (gdi2290) [inline index.js]: you also want to provide a way for devs to include other decorators such as ionic `@App` or `@Page` and provide the previous regexp if they want it to be global

File: index.js [hunk 2]
```diff
@@ -1,31 +1,37 @@
...
+          // with: template: require('./path/to/template.html')
+          return 'template:' + replaceStringsWithRequires(url);
+        })
+        .replace(stylesRegex, function (match, urls) {
+          // replace: stylesUrl: ['./foo.css', "./baz.css", "./index.component.css"]
+          // with: styles: [require('./foo.css'), require("./baz.css"), require("./index.component.css")]
+          return 'styles:' + replaceStringsWithRequires(urls);
+        });
+
+    return '@Component({' + newMetadata + '})';
```
Comment (gdi2290) [inline index.js]: you can create a group for `Component` so you can also support `@App` from ionic2

**Comment by gdi2290** [comment]
I would merge this and add the features I mentioned in the comments in another PR

**Comment by TheLarkInn** [comment]
I'm fine with this being included in this PR, if and only if we add a few more tests to account for `@App` or `@Page` etc.

**Comment by cebor** [comment]
Tests added and commits squashed.

@gdi2290 don't know if the imports for Page and App in tests are correct.

**Comment by mirkonasato** [comment]
Ionic doesn't use `@App` and `@Page` any more, since [2.0.0-beta.8](https://github.com/driftyco/ionic/blob/master/CHANGELOG.md#200-beta8-2016-06-06).

**Comment by cebor** [comment]
ok, removed @Page and @App again, also updated Regex a little bit.

**Comment by TheLarkInn** [comment]
Thank you guys. Can you shoot me you npm names

**Comment by cebor** [comment]
How do you mean?

**Comment by gdi2290** [comment]
there were a few problems with this PR's RegExp that I found awhile ago. I used it when forking this project for angular2-conventions-loader
https://github.com/AngularClass/angular2-conventions-loader/blob/master/index.js

**Pull request closed without merging**